### DISCUSSION TITLE
ESP Web Tool

### DISCUSSION THREAD

**Discussion by geekykayaker** [micropython/micropython#13222] [ESP32]
For ESP devices, ESP Home created this web tool.
It can flash ESP devices from the browser, just plug into a usb port and hit “flash”
I have used this several times to install Tasmota, and it’s incredible.
I’d love to see this added to micropython.

**Comment by geekykayaker** [comment]
I should be able to make the relevant changes myself, but want to check a few things with the Micropython devs. 

My questions are:

- What are the general thoughts on this? 
  - Any objections?
  - Any legwork I need to do?
- This is ESP specific and is slightly orthogonal to Micropython itself, should it live in the main repo? Or should maybe we create an “Installers” repo?
- Tasmota use github pages for this, is there anything I should know about Micropython team’s view/usage of these?
  - Do they have any already? 
  - Any objection to me creating one for this?

I haven’t yet checked the developers guidelines, I’ll do that next.
I also want to check the licences are comparable.

**Comment by geekykayaker** [comment]
I’ve read through the developers guide, and can’t immediately see answers to the above.

The web installer is MIT so that’s fine.
 (The core of ESP Home is GPL, so a little caution is warranted)

This repo given as an example is basically what would need to be added:  https://github.com/balloob/squeezelite-esp32-install

**Comment by dhalbert** [comment]
We (Adafruit) also have a tool like this, that I think might work fine for MicroPython.
https://adafruit.github.io/Adafruit_WebSerial_ESPTool/
https://github.com/adafruit/Adafruit_WebSerial_ESPTool

**Comment by mattytrentini** [comment]
It's certainly something worth investing time into. I've also looked into [Bipes](https://rafaelaroca.wordpress.com/2021/12/16/install-micropython-from-the-web-browser/) and Espressif's [esptool-js](https://github.com/espressif/esptool-js/) - but as @dhalbert mentioned, Adafruit WebSerial is a great starting point.

Personally, I'd like to have _all_ of the various port's firmware deployable from the browser but the ESP32 should be a great first-option!

One big caveat for me is that this is only possible on Chrome-based browsers. Mozilla ruled out the necessary WebUSB/Serial APIs as being too insecure.

**Comment by geekykayaker** [comment]
OK, so looking at these options:

- Bipes have basically done exactly what I originally suggested - using esp-web-tools
- AdaFruit WebSerial tool is a fork of esptool-js so these are kinda the same as each other

The main reason I prefer the first is that it's a single purpose tool for flashing firmwares.
It's all set up just to point at a manifest and go.
https://esphome.github.io/esp-web-tools/
Explains exactly what to do, and gives an example, as well as a list of open source projects with other examples in.
It looks like all I need to do is set up a manifest.json and it should just work.
A couple of notes on BIPES specifically is that I can't see the source for their web flasher anywhere, and their licence is GPL anyway, so that may not be suitable.

I can't really find much documentation on how to create a flashing page with AdaFruit/esptool
The AdaFruit one links to two examples, one is dead. The other is a guide on how to use Adafruit IO website, which doesn't seem like what we want.
The esptool-js one does include a single example, but I'd have to reverse engineer it, as I'm not really familiar with TypeScript.

**Comment by DavesCodeMusings** [comment]
On a side note, I have a [VS Code extension](https://marketplace.visualstudio.com/publishers/DavesCodeMusings) that enables point and click flashing of ESPs. It's basically a wrapper around the esptool Python module. If you're a VS Code user, it's a handy way to do all your tasks from the same user interface. Just right-click and go.

![image](https://github.com/micropython/micropython/assets/61114342/edc41a55-06ef-456a-9e66-877419bb8b02)
